igvm/hypervisor_preferences.py

```diff
@@ -66,209 +96,336 @@ def __repr__(self):
     """Check if the environment of the hypervisor fits with the VM env.
 
     Make any hypervisor less likely chosen, which would have a different
     environment.
     """
-    def __init__(self, hv_env: str):
-        self.hv_env = hv_env
 
-    def __repr__(self):
+    def __init__(self, hv_env: str) -> None:
+        self.hv_env: str = hv_env
+
+    def __repr__(self) -> str:
         args = repr(self.hv_env)
 
         return '{}({})'.format(type(self).__name__, args)
 
-    def __call__(self, vm, hv) -> bool:
+    def get_score(self, vm, hv) -> Union[float, bool]:
         hypervisor_env = hv.dataset_obj[self.hv_env]
         vm_env = vm.dataset_obj['environment']
 
+        # If the environment is matching, this is our candidate!
         if hypervisor_env == vm_env:
-            return True
+            return 1.
+
+        # Unfortunately we don't always have enough HVs to ensure the same
+        # environment. Therefore we will just highly discourage the hv
+        # instead of excluding it.
+        return 0.01
 
-        return False
 
+class OverAllocation(HypervisorPreference):
+    """Check for an attribute being over-allocated than the current one."""
 
-class OverAllocation(object):
-    """Check for an attribute being over allocated than the current one"""
-    def __init__(self, attribute):
+    def __init__(self, attribute) -> None:
         self.attribute = attribute
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         args = repr(self.attribute)
 
         return '{}({})'.format(type(self).__name__, args)
 
-    def __call__(self, vm, hv):
-        # New VM has no hypervisor attribute yet.
-        if not vm.hypervisor:
-            return False
-
+    def get_score(self, vm, hv) -> Union[float, bool]:
+        # Calculate the current HVs overbooking "level".
         cur_hv_cpus = sum(
             v[self.attribute] for v in vm.hypervisor.dataset_obj['vms']
         )
         cur_hv_rl_cpus = vm.hypervisor.dataset_obj[self.attribute]
         cur_ovr_allc = float(cur_hv_cpus) / float(cur_hv_rl_cpus)
 
+        # Calculate by how much we would overbook the target HV.
         tgt_hv_cpus = vm.dataset_obj[self.attribute] + sum(
             v[self.attribute] for v in hv.dataset_obj['vms']
         )
         tgt_hv_rl_cpus = hv.dataset_obj[self.attribute]
         tgt_ovr_allc = float(tgt_hv_cpus) / float(tgt_hv_rl_cpus)
 
-        return tgt_ovr_allc > cur_ovr_allc
-
+        # Whether the target hv would be more overbooked than the current one.
+        rel_overbooking = tgt_ovr_allc / cur_ovr_allc
+        if rel_overbooking > 1.:
+            return .01
 
-class HashDifference(object):
-    """Return some arbitrary number to have stable ordering"""
-    def __repr__(self):
-        return '{}()'.format(type(self).__name__)
+        # Normalize the value. This is no hard criteria, except for non-
+        # overbookable resources like disk. Hence we can allow overbooked cpus.
```

Comment:
This is true, but the logic is opposite. There is hard criteria for majority of resources (memory and storage pool), the CPU is the exception.

Reply:
changed the comment. please check it again